### PR TITLE
Fix iAj and use pj in bin.c ##refactor

### PR DESCRIPTION
**Pull request by eagleoflqj**
**Checklist**

- [ ] Closing issues: #issue
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

https://github.com/rizinorg/rizin/issues/200
Before:
```sh
liumeo@liumeo-x64:~/radare2/test$ r2 -Qc iAj bins/mach0/fatmach0-3true
"bins":[{["arch":"x86","bits":64,"offset":4096,"size":13792,"machine":"x86 64 all"]}
{["arch":"x86","bits":32,"offset":20480,"size":13760,"machine":"386"]}
{["arch":"ppc","bits":32,"offset":36864,"size":13616,"machine":"7400"]}
]liumeo@liumeo-x64:~/radare2/test$
```
Now:
```sh
liumeo@liumeo-imac:~/github/radare2/test$ r2 -Qc iAj bins/mach0/fatmach0-3true
{"bins":[{"arch":"x86","bits":64,"offset":4096,"size":13792,"machine":"x86 64 all"},{"arch":"x86","bits":32,"offset":20480,"size":13760,"machine":"386"},{"arch":"ppc","bits":32,"offset":36864,"size":13616,"machine":"7400"}]}
```
Both `rabin2 -Aj` and `r2 -Qc iAj` are fixed. See tests.
Moreover:
* Remove `r_bin_file_find_by_name_n` because it's useless after 1fbc7f21683e481a4c6ca6ba6b30a80117418b57
* Fix wrong pj imported #17636 
* Fix wrong r_table format
* Remove some useless newlines

### CODE REVIEW

**Review by trufae — APPROVED** [review]

File: libr/bin/bin.c [hunk 1]
```diff
@@ -1139,8 +1148,8 @@ R_API void r_bin_list_archs(RBin *bin, int mode) {
 				pj_end (pj);
 				break;
 			default:
-				r_table_add_rowf (table, "nsnss", i, sdb_fmt ("0x%08" PFMT64x , boffset), obj_size, "", "");
-				bin->cb_printf ("%s\n", r_table_tostring(table));
+				r_table_add_rowf (table, fmt, 0, boffset, obj_size, "", "");
+				bin->cb_printf ("%s", r_table_tostring(table));
```
Comment (trufae) [inline libr/bin/bin.c:1152]: missing space

File: libr/core/cmd_info.c [hunk 2]
```diff
@@ -595,10 +595,14 @@ static int cmd_info(void *data, const char *input) {
 				if (z) { playMsg (core, n, z);}\
 				r_core_bin_info (core, x, mode, va, NULL, y);
 		case 'A': // "iA"
-			newline = false;
-			{
-				int mode = (input[1] == 'j')? 'j': 1;
-				r_bin_list_archs (core->bin, mode);
+			if (input[1] == 'j') {
+				r_cons_print ("{");
```
Comment (trufae) [inline libr/core/cmd_info.c:599]: uhm, use pj_o inside r_bin_list_archs instead of printing the braces here

File: libr/bin/bin.c [hunk 3]
```diff
@@ -997,53 +1009,51 @@ static void list_xtr_archs(RBin *bin, int mode) {
...
-		r_table_free (table);
 		return;
 	}
-	i = -1;
+	RTable *table = r_table_new ();
+	const char *fmt = "dXnss";
+	PJ *pj;
+	if (mode == 'j') {
+		pj = pj_new ();
+		if (!pj) {
```
Comment (trufae) [inline libr/bin/bin.c:1052]: add pj_o(pj);

File: libr/bin/bin.c [hunk 4]
```diff
@@ -1151,14 +1160,13 @@ R_API void r_bin_list_archs(RBin *bin, int mode) {
 		//sdb_array_push (binfile_sdb, ARCHS_KEY, archline, 0);
 	}
 	if (mode == 'j') {
-		pj_end (pj);
 		pj_end (pj);
```
Comment (trufae) [inline libr/bin/bin.c:1163]: add one more pj_end

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_info.c (hunk 2 above)
Comment (eagleoflqj) [reply libr/core/cmd_info.c:599]: That will break `rabin2 -Aj`.
https://github.com/radareorg/radare2/blob/38391b38cb11b7b394b3cc4b2abc6c408dab4857/libr/main/rabin2.c#L1111-L1124
See line 1112. Here the call to r_bin_list_archs is wrapped in manually printed "{" and "}". Afterwards there are potentially other keys in this json object.
For example `rabin2 -Aij bins/mach0/fatmach0-3true` prints
```
{"bins":[{"arch":"x86","bits":64,"offset":4096,"size":13792,"machine":"x86 64 all"},{"arch":"x86","bits":32,"offset":20480,"size":13760,"machine":"386"},{"arch":"ppc","bits":32,"offset":36864,"size":13616,"machine":"7400"}],"imports":[{"ordinal":0,"bind":"NONE","type":"FUNC","name":"exit","plt":4294971200},{"ordinal":1,"bind":"NONE","type":"FUNC","name":"dyld_stub_binder","plt":0}]}
```
So your suggestion can only be applied if we pass pj as argument to `r_bin_list_archs`. And also more refactor are needed in rabin2.c.

File: libr/bin/bin.c (hunk 3 above)
Comment (eagleoflqj) [reply libr/bin/bin.c:1052]: Same with above.

File: libr/bin/bin.c (hunk 4 above)
Comment (eagleoflqj) [reply libr/bin/bin.c:1163]: Same.

File: libr/core/cmd_info.c (hunk 2 above)
Comment (trufae) [reply libr/core/cmd_info.c:599]: Ok dont overcomplicate this PR for now :) thanks . The proper fix will require more refactoring as you say

**Pull request merged**